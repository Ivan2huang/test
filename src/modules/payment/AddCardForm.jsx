import React, { useEffect, useState, memo, useCallback } from 'react';

import PropTypes from 'prop-types';
import { injectIntl } from 'react-intl';
import { compose } from 'redux';

import { Box, Button } from '@material-ui/core';
import Radio from '@material-ui/core/Radio';
import FormControl from '@material-ui/core/FormControl';
import RadioGroup from '@material-ui/core/RadioGroup';
import FormControlLabel from '@material-ui/core/FormControlLabel';

import Typography from '../../uiComponents/Typography';
import AddNewCard from './AddNewCard';
import { instrumentType } from '../../types';
import { formatMessage } from '../../helpers/helpers';
import { ADD_NEW_CARD } from './constants';
import { IntlPropType } from '../../i18n/lang';

const AddCardForm = ({
  intl,
  instruments,
  preAuthInstrument,
  submitEnrollInstrument,
  isLoadingInstruments,
}) => {
  const [state, setState] = useState(0);

  useEffect(() => {
    setState(`${instruments[0].id}`);
  }, [instruments, setState, isLoadingInstruments]);

  const onChangeSession = useCallback(e => {
    setState(e.target.value);
  }, []);

  const onPreAuthInstrument = () => {
    preAuthInstrument(state);
  };

  const formatCardNumber = number => {
    const lastNumber = number.slice(-4);
    let maskedNumber = '';

    for (let i = 0; i < number.length - 4; i += 1) {
      maskedNumber += i % 4 === 3 ? '* ' : '*';
    }

    return `${maskedNumber} ${lastNumber}`;
  };

  return (
    <>
      <Typography type="style-6">
        {formatMessage(
          intl,
          'payment.paymentForm.information.description',
          'To verify the validity of your existing payment option, SGD$1.00 will be charged and refunded automatically for pre-authorisation of your credit card.',
        )}
      </Typography>

      <Box display="flex" justifyContent="flex-start" mt={4}>
        <Typography type="style-6">
          {formatMessage(
            intl,
            'payment.paymentForm.information.subDescription',
            'The total consultation fees will be sent to you after your consultation.',
          )}
        </Typography>
      </Box>

      <Box display="flex" justifyContent="flex-start" mt={12}>
        <Typography type="style-5" fontWeight="bold">
          {formatMessage(
            intl,
            'payment.paymentForm.paymentMethod.label',
            'Payment method',
          )}
        </Typography>
      </Box>

      {instruments.length > 1 && (
        <FormControl component="fieldset">
          <RadioGroup
            name="select-card"
            value={state}
            onChange={onChangeSession}
          >
            {instruments.map(item => (
              <FormControlLabel
                key={`${item.id}`}
                value={`${item.id}`}
                control={<Radio color="secondary" />}
                label={
                  item.id !== ADD_NEW_CARD
                    ? formatCardNumber(item.card.maskedCreditCardNumber)
                    : item.card.maskedCreditCardNumber
                }
              />
            ))}
          </RadioGroup>
        </FormControl>
      )}

      {instruments.length > 1 && state !== ADD_NEW_CARD && (
        <Box display="flex" justifyContent="flex-start" mt={14}>
          <Button
            type="submit"
            data-testid="btn-submit-preauth"
            color="primary"
            variant="contained"
            onClick={onPreAuthInstrument}
          >
            {formatMessage(intl, 'payment.paymentForm.button.next', 'Next')}
          </Button>
        </Box>
      )}

      <AddNewCard
        submitEnrollInstrument={submitEnrollInstrument}
        show={state === ADD_NEW_CARD && !isLoadingInstruments}
      />
    </>
  );
};

AddCardForm.propTypes = {
  instruments: PropTypes.arrayOf(instrumentType).isRequired,
  intl: IntlPropType.isRequired,
  preAuthInstrument: PropTypes.func.isRequired,
  submitEnrollInstrument: PropTypes.func.isRequired,
  isLoadingInstruments: PropTypes.bool.isRequired,
};

export default compose(
  injectIntl,
  memo,
)(AddCardForm);
