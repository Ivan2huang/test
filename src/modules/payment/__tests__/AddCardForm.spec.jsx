import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import withIntl from '../../../i18n/withIntlProvider';
import AddCardForm from '../AddCardForm';

// eslint-disable-next-line react/prop-types
jest.mock('../../../uiComponents/Typography', () => ({ children }) => {
  return <div>{children}</div>;
});

// eslint-disable-next-line react/prop-types
jest.mock('../AddNewCard', () => ({ submitEnrollInstrument }) => {
  return (
    // eslint-disable-next-line react/button-has-type
    <button
      data-testid="mock-submit-newcard"
      onClick={() => submitEnrollInstrument('sessionId')}
    />
  );
});

describe('AddCardForm Component', () => {
  let Component;
  const preAuthInstrument = jest.fn();
  const submitEnrollInstrument = jest.fn();
  const props = {
    instruments: [
      {
        id: 1,
        card: {
          maskedCreditCardNumber: 'xxxxxxxxxxx1234',
        },
      },
      {
        id: 'ADD_NEW_CARD',
        card: {
          maskedCreditCardNumber: 'New Card',
        },
      },
    ],
    preAuthInstrument,
    submitEnrollInstrument,
    isLoadingInstruments: false,
  };

  beforeEach(() => {
    Component = withIntl(AddCardForm);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match to snapshot when there is  at least one instrument', () => {
    const { container } = render(<Component {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('should match to snapshot when there is no instruments', () => {
    const { container } = render(
      <Component
        {...{
          ...props,
          instruments: [
            {
              id: 'ADD_NEW_CARD',
              card: {
                maskedCreditCardNumber: 'New Card',
              },
            },
          ],
        }}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should change selected instrument when click on radio button', () => {
    const { getByLabelText } = render(<Component {...props} />);

    const addNewRadioNode = getByLabelText(/New Card/i);
    const card1RadioNode = getByLabelText(/1234$/i);

    expect(addNewRadioNode.value).toEqual('ADD_NEW_CARD');
    expect(card1RadioNode.value).toEqual('1');

    fireEvent.click(card1RadioNode);
    expect(card1RadioNode.checked).toEqual(true);
    expect(addNewRadioNode.checked).toEqual(false);

    fireEvent.click(addNewRadioNode);
    expect(addNewRadioNode.checked).toEqual(true);
    expect(card1RadioNode.checked).toEqual(false);
  });

  it('should submit the corrent info when submit by existing instrument', () => {
    const { getByTestId, getByLabelText } = render(<Component {...props} />);

    const card1RadioNode = getByLabelText(/1234$/i);
    fireEvent.click(card1RadioNode);
    const submitCheckout = getByTestId('btn-submit-preauth');
    fireEvent.click(submitCheckout);

    expect(preAuthInstrument).toBeCalledTimes(1);
    expect(preAuthInstrument).toBeCalledWith('1');
  });

  it('should checkout by new card correctly', () => {
    const { getByLabelText, getByTestId } = render(<Component {...props} />);

    const addNewRadioNode = getByLabelText(/New Card/i);
    fireEvent.click(addNewRadioNode);

    const submit = getByTestId('mock-submit-newcard');
    fireEvent.click(submit);

    expect(submitEnrollInstrument).toHaveBeenCalledTimes(1);
    expect(submitEnrollInstrument).toHaveBeenCalledWith('sessionId');
  });
});
