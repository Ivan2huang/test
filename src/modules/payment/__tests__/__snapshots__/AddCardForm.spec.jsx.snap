// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCardForm Component should match to snapshot when there is  at least one instrument 1`] = `
<div>
  <div>
    To verify the validity of your existing payment option, SGD$1.00 will be charged and refunded automatically for pre-authorisation of your credit card.
  </div>
  <div
    class="MuiBox-root MuiBox-root-2"
  >
    <div>
      The total consultation fees will be sent to you after your consultation.
    </div>
  </div>
  <div
    class="MuiBox-root MuiBox-root-3"
  >
    <div>
      Payment method
    </div>
  </div>
  <fieldset
    class="MuiFormControl-root"
  >
    <div
      class="MuiFormGroup-root"
      role="radiogroup"
    >
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-21 MuiRadio-root MuiRadio-colorSecondary PrivateSwitchBase-checked-22 Mui-checked MuiIconButton-colorSecondary"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-24"
              name="select-card"
              type="radio"
              value="1"
            />
            <div
              class="PrivateRadioButtonIcon-root-37 PrivateRadioButtonIcon-checked-39"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-38"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                />
              </svg>
            </div>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          **** **** *** 1234
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-21 MuiRadio-root MuiRadio-colorSecondary MuiIconButton-colorSecondary"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-24"
              name="select-card"
              type="radio"
              value="ADD_NEW_CARD"
            />
            <div
              class="PrivateRadioButtonIcon-root-37"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-38"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                />
              </svg>
            </div>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          New Card
        </span>
      </label>
    </div>
  </fieldset>
  <div
    class="MuiBox-root MuiBox-root-79"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      data-testid="btn-submit-preauth"
      tabindex="0"
      type="submit"
    >
      <span
        class="MuiButton-label"
      >
        Next
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
  <button
    data-testid="mock-submit-newcard"
  />
</div>
`;

exports[`AddCardForm Component should match to snapshot when there is no instruments 1`] = `
<div>
  <div>
    To verify the validity of your existing payment option, SGD$1.00 will be charged and refunded automatically for pre-authorisation of your credit card.
  </div>
  <div
    class="MuiBox-root MuiBox-root-119"
  >
    <div>
      The total consultation fees will be sent to you after your consultation.
    </div>
  </div>
  <div
    class="MuiBox-root MuiBox-root-120"
  >
    <div>
      Payment method
    </div>
  </div>
  <button
    data-testid="mock-submit-newcard"
  />
</div>
`;
